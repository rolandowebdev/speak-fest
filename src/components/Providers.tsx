'use client'
import { useRef } from 'react'
import { Provider } from 'react-redux'
import { makeStore, AppStore } from '../lib/store'

export default function StoreProvider({
	children,
}: {
	children: React.ReactNode
}) {
	const storeRef = useRef<AppStore | null>(null)
	if (!storeRef.current) {
		storeRef.current = makeStore()
	}

	return <Provider store={storeRef.current}>{children}</Provider>
}
