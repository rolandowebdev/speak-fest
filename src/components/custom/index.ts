export * from './button'
export * from './header'
