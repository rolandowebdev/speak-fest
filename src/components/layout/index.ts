export * from './footer'
export * from './navigation'
export * from './page-container'
