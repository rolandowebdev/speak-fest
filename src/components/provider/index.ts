export * from './redux'
export * from './theme'
