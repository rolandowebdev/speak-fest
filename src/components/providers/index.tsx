'use client'

import * as React from 'react'
import { SessionProvider } from 'next-auth/react'
import StoreProvider from '@/libs/redux/store-provider'
import ThemeProvider from './theme'

type ProvidersProps = {
  children: React.ReactNode
}

export default function Providers({ children }: ProvidersProps) {
  return (
    <StoreProvider>
      <ThemeProvider
        attribute="class"
        defaultTheme="system"
        enableSystem
        disableTransitionOnChange>
        <SessionProvider>{children}</SessionProvider>
      </ThemeProvider>
    </StoreProvider>
  )
}
