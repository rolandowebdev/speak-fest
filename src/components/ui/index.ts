export * from './button'
export * from './heading'
export * from './sheet'
export * from './form'
export * from './input'
export * from './label'
export * from './avatar'
export * from './skeleton'
