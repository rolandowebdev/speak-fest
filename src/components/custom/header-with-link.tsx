import * as React from 'react'
import { Button, Heading } from '@/components/ui'
import { Undo2 } from 'lucide-react'
import { useRouter } from 'next/navigation'
import { Header } from '@/components/layout'

type HeaderWithLinkProps = {
  icon: React.ReactNode
  title: string
}

export function HeaderWithLink({ icon, title }: HeaderWithLinkProps) {
  const { push } = useRouter()

  return (
    <Header>
      <Button
        data-testid="back-to-home"
        variant="link"
        className="flex items-center gap-1 px-0 text-lg text-primary"
        onClick={() => push('/')}>
        <Undo2 size={18} />
        Back to home
      </Button>
      <Heading className="flex flex-wrap items-center gap-2">
        {icon} {title}
      </Heading>
    </Header>
  )
}
