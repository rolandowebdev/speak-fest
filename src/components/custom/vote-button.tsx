import * as React from 'react'
import { Button } from '@/components/ui'
import { ThumbsDown, ThumbsUp } from 'lucide-react'

export type VoteButtonProps = {
  voteType: 'up-vote' | 'down-vote'
  isVoted: boolean
  voteCount: number
  onClick: () => void
  authStatus: 'authenticated' | 'unauthenticated' | 'loading' | undefined
}

export function VoteButton({
  voteType,
  isVoted,
  voteCount,
  onClick,
  authStatus,
}: VoteButtonProps) {
  let icon

  if (voteType === 'up-vote') {
    icon = isVoted ? (
      <ThumbsUp
        size={18}
        className="fill-blue-500 dark:fill-blue-600"
        data-testid="thumbs-up-icon"
      />
    ) : (
      <ThumbsUp size={18} data-testid="thumbs-up-icon" />
    )
  } else {
    icon = isVoted ? (
      <ThumbsDown
        size={18}
        className="fill-blue-500 dark:fill-blue-600"
        data-testid="thumbs-down-icon"
      />
    ) : (
      <ThumbsDown size={18} data-testid="thumbs-down-icon" />
    )
  }

  return (
    <Button
      variant="ghost"
      onClick={onClick}
      className="flex items-center gap-1 text-lg"
      disabled={authStatus === 'unauthenticated' || authStatus === 'loading'}>
      {icon}
      <span className="text-lg">{voteCount || 0}</span>
    </Button>
  )
}
