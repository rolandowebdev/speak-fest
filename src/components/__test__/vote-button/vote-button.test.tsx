/**
 * Scenario test for VoteButton component :
 * - should render correctly icon based on up-vote type
 * - should render correctly icon based on down-vote type
 * - should disable vote button when auth status is loading
 * - should disable vote button when auth status is unauthenticated
 * - should undisable vote button when auth status is authenticated
 */

import * as React from 'react'
import { describe, it } from '@jest/globals'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { VoteButton, VoteButtonProps } from '@/components/custom'

describe('VoteButton component', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  const defaultProps: VoteButtonProps = {
    voteType: 'up-vote',
    isVoted: false,
    voteCount: 0,
    onClick: jest.fn(),
    authStatus: 'authenticated',
  }

  it('should render correctly icon based on vote type for up-vote', () => {
    render(<VoteButton {...defaultProps} voteType="up-vote" />)
    const voteButton = screen.getByRole('button')
    const thumbsUpIcon = screen.getByTestId('thumbs-up-icon')

    expect(voteButton).toMatchSnapshot()
    expect(thumbsUpIcon).toMatchSnapshot()
  })

  it('should render correctly icon based on vote type for down-vote', () => {
    render(<VoteButton {...defaultProps} voteType="down-vote" />)
    const voteButton = screen.getByRole('button')
    const thumbsDownIcon = screen.getByTestId('thumbs-down-icon')

    expect(voteButton).toMatchSnapshot()
    expect(thumbsDownIcon).toMatchSnapshot()
  })

  it('should disable vote button when auth status is loading', () => {
    render(<VoteButton {...defaultProps} authStatus="loading" />)
    const voteButtonLoading = screen.getByRole('button')
    expect(voteButtonLoading).toBeDisabled()
    expect(voteButtonLoading).toMatchSnapshot()
  })

  it('should disable vote button when auth status is unauthenticated', () => {
    render(<VoteButton {...defaultProps} authStatus="unauthenticated" />)
    const voteButtonUnauthenticated = screen.getByRole('button')
    expect(voteButtonUnauthenticated).toBeDisabled()
    expect(voteButtonUnauthenticated).toMatchSnapshot()
  })

  it('should undisable vote button when auth status is authenticated', () => {
    render(<VoteButton {...defaultProps} authStatus="authenticated" />)
    const voteButtonAuthenticated = screen.getByRole('button')
    expect(voteButtonAuthenticated).not.toBeDisabled()
    expect(voteButtonAuthenticated).toMatchSnapshot()
  })
})
