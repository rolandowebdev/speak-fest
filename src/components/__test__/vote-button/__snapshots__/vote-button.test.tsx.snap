// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteButton component should disable vote button when auth status is loading 1`] = `
<button
  class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center gap-1 text-lg"
  disabled=""
>
  <svg
    class="lucide lucide-thumbs-up "
    data-testid="thumbs-up-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10v12"
    />
    <path
      d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
    />
  </svg>
  <span
    class="text-lg"
  >
    0
  </span>
</button>
`;

exports[`VoteButton component should disable vote button when auth status is unauthenticated 1`] = `
<button
  class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center gap-1 text-lg"
  disabled=""
>
  <svg
    class="lucide lucide-thumbs-up "
    data-testid="thumbs-up-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10v12"
    />
    <path
      d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
    />
  </svg>
  <span
    class="text-lg"
  >
    0
  </span>
</button>
`;

exports[`VoteButton component should render correctly icon based on vote type for down-vote 1`] = `
<button
  class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center gap-1 text-lg"
>
  <svg
    class="lucide lucide-thumbs-down "
    data-testid="thumbs-down-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17 14V2"
    />
    <path
      d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22h0a3.13 3.13 0 0 1-3-3.88Z"
    />
  </svg>
  <span
    class="text-lg"
  >
    0
  </span>
</button>
`;

exports[`VoteButton component should render correctly icon based on vote type for down-vote 2`] = `
<svg
  class="lucide lucide-thumbs-down "
  data-testid="thumbs-down-icon"
  fill="none"
  height="18"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="18"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M17 14V2"
  />
  <path
    d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22h0a3.13 3.13 0 0 1-3-3.88Z"
  />
</svg>
`;

exports[`VoteButton component should render correctly icon based on vote type for up-vote 1`] = `
<button
  class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center gap-1 text-lg"
>
  <svg
    class="lucide lucide-thumbs-up "
    data-testid="thumbs-up-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10v12"
    />
    <path
      d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
    />
  </svg>
  <span
    class="text-lg"
  >
    0
  </span>
</button>
`;

exports[`VoteButton component should render correctly icon based on vote type for up-vote 2`] = `
<svg
  class="lucide lucide-thumbs-up "
  data-testid="thumbs-up-icon"
  fill="none"
  height="18"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="18"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M7 10v12"
  />
  <path
    d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
  />
</svg>
`;

exports[`VoteButton component should undisable vote button when auth status is authenticated 1`] = `
<button
  class="justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center gap-1 text-lg"
>
  <svg
    class="lucide lucide-thumbs-up "
    data-testid="thumbs-up-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 10v12"
    />
    <path
      d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
    />
  </svg>
  <span
    class="text-lg"
  >
    0
  </span>
</button>
`;
